Add E2E tests for Azure internal loadbalancer support, fix an issue for public IP resource deletion.

karataliu opened the pull request:
**What this PR does / why we need it**:

- Add E2E tests for Azure internal loadbalancer support: https://github.com/kubernetes/kubernetes/pull/43510
- Fix an issue that public IP resource not get deleted when switching from external loadbalancer to internal static loadbalancer.

**Special notes for your reviewer**:

1.  Add new Azure resource tag to Public IP resources to indicate kubernetes managed resources.
   Currently we determine whether the public IP resource should be deleted by looking at LoadBalancerIp property on spec. In the scenario 'Switching from external loadbalancer to internal loadbalancer with static IP', that value might have been updated for internal loadbalancer. So here we're to add an explicit tag for kubernetes managed resources.

2. Merge cleanupPublicIP logic into cleanupLoadBalancer

**Release note**:
NONE

CC @brendandburns @colemickens

karataliu committed:
Add E2E tests for Azure internal loadbalancer support, fix an issue for public IP resource deletion.

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/45473)
<!-- Reviewable:end -->

spiffxp commented:
@k8s-bot ok to test

justinsb commented:
/assign @colemickens for triage

/unassign @justinsb
/unassign @saad-ali

colemickens commented:
@karataliu I can understand that there would be a bug about orphaning a public IP when switching from External->Internal, but I don't understand what the tag is for... we should only ever be looking at resources that are prefixed with the service UUID... and that value doesn't change unless the Service is deleted and re-created, in which case the ServiceController is going to delete that IP config, etc, and then recreate it like new (except for the other edge case caused by ServiceController tracking these things by name, rather than UUID... but that's a separate issue). I'm confused about why we would ever be considering other resources such that we need to evaluate that tag.

karataliu commented:
Thanks for the comment.

A previous change has removed 'getPublicIPName' from util, and add one in azure_loadbalancer.go to figure out the potential user public IP resource:
https://github.com/kubernetes/kubernetes/commit/ea23cabfa05e8a1ae5badb9dafae326510d318cf#diff-982393cb8bac49b84e87ca8807d4e5e8L195 

I've updated this PR accordingly:
1. Add the original getPulicIPName back to util, and rename the current one in azuer_loadbalancer.go to determinePublicIPName.
2. Updated to use name comparison to determine public IP keeping logic to keep consistent.

jdumars commented:
/sig azure

jdumars commented:
/retest

colemickens commented:
@karataliu This LGTM. Do you plan to document this on kubernetes.github.io repo? It would be good, if you do, if you can mention that the PIP resource has to be in the same resource group as the cluster? (Unless I missed it, this all still looks to be oriented around resources "names" all in the context of a single RG.)

/lgtm

colemickens commented:
/assign @ixdy

ixdy commented on test/e2e/service.go:
```diff
@@ -1241,6 +1241,77 @@ var _ = framework.KubeDescribe("Services", func() {
 		framework.CheckReachabilityFromPod(true, normalReachabilityTimeout, namespace, dropPodName, svcIP)
 	})
+
+	It("should be able to create an internal type load balancer [Slow]", func() {
```
maybe put Azure in the name, since this is an Azure-only test?

karataliu commented:
@ixdy update pushed.

karataliu commented:
@colemickens Yes, current implementation would require the public IP resource to be in the same resource group of the cluster. I'm to prepare the doc update PR later.

karataliu commented:
@colemickens  Please help review: https://github.com/kubernetes/kubernetes.github.io/pull/4082

ixdy commented:
/approve

Thanks! Can you please squash your commits, too?

ixdy commented:
I don't think e2e tests need a tracking issue.
/approve no-issue

karataliu commented:
squash commit pushed. thanks.

ixdy commented:
/lgtm
/release-note-none

ixdy commented:
sorry for delay; should this go into 1.7?

karataliu commented:
Thanks for update. I'd suggest making this in v1.7 . The previous PR#43510 for Azure internal LoadBalancer support is new in v1.7, and this one brings e2e case and bug fix.

ixdy commented:
ok, tagging it as v1.7. @kubernetes/kubernetes-release-managers @dchen1107

ixdy commented:
@karataliu is there an issue associated with the bug this is fixing?

karataliu commented:
@ixdy no issue created yet, the origin feature issue is here: #38901

Shall I create one? Recap the bug description here based on fixups:


For external LoadBalancer feature on Azure, if a user exposes a Kubernetes service with LoadBalancer type, a new public IP resource will be created. And when the service got deleted, the public IP resource should also be deleted.

In #42034, we supported user specified IP for LoadBalancer on Azure, that is, a user could provision a public IP resource at first, and then specify it as the `LoadBalancerIP` value in service spec. In this case, when the service got deleted, it will not delete the user provided public IP. 

Currently the logic for determining whether the public IP resource should be deleted is by checking whether `LoadBalancerIP` value in spec is empty. This becomes a issue after we later introduced internal LoadBalancer support in #43510. When a user switches from external LoadBalancer to internal LoadBalancer with static IP, we are not able to tell whether the public IP resource should be deleted, since the `LoadBalancerIP` value is overrided by the internal LoadBalancer's service spec.

The key thing is how to identify a public IP resource is _managed_ or not. One solution could be determine by the name, since the managed public IPs will have conventional naming, which includes cluster name and service uuid.
